src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs

```diff
@@ -0,0 +1,420 @@
+
+                for (int i = 0; i < entryCount; ++i)
+                {
+                    AndroidTzDataEntry entry = LoadEntryAt(fs, indexOffset + (entrySize*i));
+                    var p = (sbyte*)entry.id;
+
+                    _byteOffsets![i] = entry.byteOffset + dataOffset;
+                    _ids![i] = new string(p);
+                    _lengths![i] = entry.length;
+
+                    if (_lengths![i] < 24) // AndroidTzDataHeader
+                    {
+                        throw new InvalidOperationException(SR.InvalidOperation_BadIndexLength);
+                    }
+                }
+            }
+
+            private void ReadTzDataIntoBuffer(Stream fs, long position, Span<byte> buffer)
+            {
+                fs.Position = position;
+
+                int bytesRead = 0;
+                int bytesLeft = buffer.Length;
+
+                while (bytesLeft > 0)
+                {
+                    int b = fs.Read(buffer.Slice(bytesRead));
+                    if (b == 0)
+                    {
+                        break;
+                    }
+
+                    bytesRead += b;
+                    bytesLeft -= b;
+                }
+
+                if (bytesLeft != 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_ReadTZError, _tzFilePath, position, buffer.Length, bytesRead, buffer.Length));
+                }
+            }
+
+            private unsafe AndroidTzDataEntry LoadEntryAt(Stream fs, long position)
+            {
+                int size = 52; // AndroidTzDataEntry
+                Span<byte> entryBuffer = stackalloc byte[size];
+
+                ReadTzDataIntoBuffer(fs, position, entryBuffer);
+
+                AndroidTzDataEntry entry;
+                for (int i = 0; i < 40; i++)
+                {
+                    entry.id[i] = entryBuffer[i];
+                }
+                entry.byteOffset = TZif_ToInt32(entryBuffer.Slice(40, 4));
+                entry.length = TZif_ToInt32(entryBuffer.Slice(44, 4));
+
+                return entry;
+            }
+
+            public string[] GetTimeZoneIds()
+            {
+                return _ids!;
+            }
+
+            public string GetTimeZoneDirectory()
+            {
+                return _tzFilePath;
+            }
+
+            public byte[] GetTimeZoneData(string id)
+            {
+                int i = Array.BinarySearch(_ids!, id, StringComparer.Ordinal);
+                if (i < 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.TimeZoneNotFound_MissingData, id));
+                }
+
+                int offset = _byteOffsets![i];
+                int length = _lengths![i];
+
+                Span<byte> buffer = stackalloc byte[length];
+                ReadTzDataIntoBuffer(File.OpenRead(_tzFilePath), offset, buffer);
+                byte[] tzBuffer = buffer.ToArray();
```

Comment:
No reason to make a stack buffer here, copy the file into the buffer, and then ToArray it.

1. It is unnecessary copying
2. We don't know that the length of the file will fit on the stack. In general, stackalloc's should not be unbounded.

```suggestion
                byte[] tzBuffer = new byte[length];
                ReadTzDataIntoBuffer(File.OpenRead(_tzFilePath), offset, tzBuffer);
```

Reply:
Thanks. I see, was it unnecessary copying because we need to still return a `byte[]` here, whereas in `LoadTzFile` we didn't "need" to use a `byte[]` and a `Span<byte>` would suffice?

Is there a general rule for how large a stackalloc should be?

Reply:
>  was it unnecessary copying because we need to still return a byte[] here, whereas in LoadTzFile we didn't "need" to use a byte[] and a Span<byte> would suffice?

Correct. In LoadTzFile, you just need a buffer to read the file data into, and then you are parsing the values out of it from the buffer. The buffer didn't need to stick around.

Here, since the method expects you to return a `byte[]`, you might as well just read directly into the `byte[]` you are going to return.

> Is there a general rule for how large a stackalloc should be?

Looking around the codebase, we usually don't let a stackalloc go above 1K of memory.